pkg/util/httprequest/httprequest.go

```diff
@@ -68,43 +68,52 @@ func SchemeHost(req *http.Request) (string /*scheme*/, string /*host*/) {
 
 	forwardedProto := forwarded("Proto")
 	forwardedHost := forwarded("Host")
-	// If both X-Forwarded-Host and X-Forwarded-Port are sent, use the explicit port info
-	if forwardedPort := forwarded("Port"); len(forwardedHost) > 0 && len(forwardedPort) > 0 {
-		if h, _, err := net.SplitHostPort(forwardedHost); err == nil {
-			forwardedHost = net.JoinHostPort(h, forwardedPort)
-		} else {
-			forwardedHost = net.JoinHostPort(forwardedHost, forwardedPort)
-		}
+	forwardedPort := forwarded("Port")
+
+	scheme := ""
+	switch {
+	case len(forwardedProto) > 0:
+		scheme = forwardedProto
+	case req.TLS != nil:
+		scheme = "https"
+	case len(req.URL.Scheme) > 0:
+		scheme = req.URL.Scheme
 	}
 
-	host := ""
+	hostHeader := ""
 	switch {
 	case len(forwardedHost) > 0:
-		host = forwardedHost
+		hostHeader = forwardedHost
 	case len(req.Host) > 0:
-		host = req.Host
+		hostHeader = req.Host
 	case len(req.URL.Host) > 0:
-		host = req.URL.Host
+		hostHeader = req.URL.Host
 	}
 
-	port := ""
-	if _, p, err := net.SplitHostPort(host); err == nil {
-		port = p
+	host, port, err := net.SplitHostPort(hostHeader)
+	if err != nil {
+		host = hostHeader
+		port = ""
 	}
 
-	scheme := ""
-	switch {
-	case len(forwardedProto) > 0:
-		scheme = forwardedProto
-	case req.TLS != nil:
-		scheme = "https"
-	case len(req.URL.Scheme) > 0:
-		scheme = req.URL.Scheme
-	case port == "443":
-		scheme = "https"
-	default:
-		scheme = "http"
+	// If X-Forwarded-Port is sent, use the explicit port info
+	if len(forwardedPort) > 0 && (len(port) != 0 || len(scheme) == 0 && forwardedPort != "80" && forwardedPort != "443") {
```

Comment:
with that configuration, this logic leaves the port as "", which would redirect to `https://example.com`, which implies port 443, which is wrong... we should send `https://example.com:80` in that case.

stepping back, can we remember whether the original host included an explicit port, and if it did not, insert a final step that strips default ports that match the scheme?

```
host := ""
hostHadExplicitPort := false
switch {
case len(forwardedHost) > 0:
	host = forwardedHost
	hostHadExplicitPort = hasExplicitPort(forwarded("Host"))
case len(req.Host) > 0:
	host = req.Host
	hostHadExplicitPort = hasExplicitPort(host)
case len(req.URL.Host) > 0:
	host = req.URL.Host
	hostHadExplicitPort = hasExplicitPort(host)
}
...

// Do not include a port if the original host did not explicitly set it, and the port is a default port that matches the scheme
if !hostHadExplicitPort {
  if (scheme == "https" && port == "443") || ("scheme" == "http" && port == "80") {
    if hostWithoutPort, _, err := net.SplitHostPort(host); err == nil {
      host = hostWithoutPort
    }
  }
}

return scheme, host
```